batch/batch/driver/gce.py

```diff
@@ -0,0 +1,147 @@
+import re
+import json
+import asyncio
+import logging
+import dateutil.parser
+import datetime
+
+from hailtop import aiotools
+
+from ..batch_configuration import PROJECT
+
+log = logging.getLogger('gce_event_monitor')
+
+RESOURCE_NAME_REGEX = re.compile('projects/(?P<project>[^/]+)/zones/(?P<zone>[^/]+)/instances/(?P<name>.+)')
+
+
+def parse_resource_name(resource_name):
+    match = RESOURCE_NAME_REGEX.fullmatch(resource_name)
+    assert match
+    return match.groupdict()
+
+
+class GCEEventMonitor:
+    def __init__(self, app, machine_name_prefix):
+        self.app = app
+        self.db = app['db']
+        self.zone_monitor = app['zone_monitor']
+        self.compute_client = app['compute_client']
+        self.logging_client = app['logging_client']
+        self.inst_coll_manager = app['inst_coll_manager']
+        self.machine_name_prefix = machine_name_prefix
+
+        self.task_manager = aiotools.BackgroundTaskManager()
+
+    async def async_init(self):
+        self.task_manager.ensure_future(self.event_loop())
+
+    def shutdown(self):
+        self.task_manager.shutdown()
+
+    async def handle_preempt_event(self, instance, timestamp):
+        await instance.inst_coll.call_delete_instance(instance, 'preempted', timestamp=timestamp)
+
+    async def handle_delete_done_event(self, instance, timestamp):
+        await instance.inst_coll.remove_instance(instance, 'deleted', timestamp)
+
+    async def handle_call_delete_event(self, instance, timestamp):
+        await instance.mark_deleted('deleted', timestamp)
+
+    async def handle_event(self, event):
+        payload = event.get('protoPayload')
+        if payload is None:
+            log.warning('event has no payload')
```

Comment:
I think whenever something goes wrong handling the event, we should include the event dict in the message. I remember debugging some issue with workers a while ago and it was quite frustrating that the bad event wasn't included in the error message.